From 6d6f2d3543cd4b172aa85f0e47d7f531b7ec4d53 Mon Sep 17 00:00:00 2001
From: Brendan Heywood <brendan@catalyst-au.net>
Date: Wed, 18 Oct 2017 16:20:33 +1100
Subject: [PATCH 1/2] MDL-60470 core: New hook 'after_require_login'

This adds a hook towards the end of the require_login function.
---
 lib/moodlelib.php | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/lib/moodlelib.php b/lib/moodlelib.php
index 3ac3d8be1b7..d9e5baa175e 100644
--- a/lib/moodlelib.php
+++ b/lib/moodlelib.php
@@ -2705,6 +2705,8 @@ function require_login($courseorid = null, $autologinguest = true, $cm = null, $
         $CFG->forceclean = true;
     }

+    $afterlogins = get_plugins_with_function('after_require_login', 'lib.php');
+
     // Do not bother admins with any formalities, except for activities pending deletion.
     if (is_siteadmin() && !($cm && $cm->deletioninprogress)) {
         // Set the global $COURSE.
@@ -2716,6 +2718,12 @@ function require_login($courseorid = null, $autologinguest = true, $cm = null, $
         }
         // Set accesstime or the user will appear offline which messes up messaging.
         user_accesstime_log($course->id);
+
+        foreach ($afterlogins as $plugintype => $plugins) {
+            foreach ($plugins as $pluginfunction) {
+                $pluginfunction($courseorid, $autologinguest, $cm, $setwantsurltome, $preventredirect);
+            }
+        }
         return;
     }

@@ -2923,6 +2931,12 @@ function require_login($courseorid = null, $autologinguest = true, $cm = null, $
         $PAGE->set_course($course);
     }

+    foreach ($afterlogins as $plugintype => $plugins) {
+        foreach ($plugins as $pluginfunction) {
+            $pluginfunction($courseorid, $autologinguest, $cm, $setwantsurltome, $preventredirect);
+        }
+    }
+
     // Finally access granted, update lastaccess times.
     user_accesstime_log($course->id);
 }
--
2.17.1


From 7235752ed449ab6662a317f059e444598bf1a862 Mon Sep 17 00:00:00 2001
From: Brendan Heywood <brendan@catalyst-au.net>
Date: Thu, 8 Aug 2019 13:26:50 +1000
Subject: [PATCH 2/2] MDL-66340 setup: Add after_config for after setup.php is
 loaded

---
 lib/setup.php | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/lib/setup.php b/lib/setup.php
index b1cb7e054ec..df1a2d978a5 100644
--- a/lib/setup.php
+++ b/lib/setup.php
@@ -1042,3 +1042,16 @@ if (false) {
     $OUTPUT = new core_renderer(null, null);
     $PAGE = new moodle_page();
 }
+
+// Allow plugins to callback as soon possible after setup.php is loaded.
+$pluginswithfunction = get_plugins_with_function('after_config', 'lib.php');
+foreach ($pluginswithfunction as $plugins) {
+    foreach ($plugins as $function) {
+        try {
+            $function();
+        } catch (Exception $e) {
+            debugging("Exception calling '$function'", DEBUG_DEVELOPER, $e->getTrace());
+        }
+    }
+}
+
--
2.17.1
